File: packages/shared/src/money.ts
/**
 * Integer money arithmetic.
 *
 * Every monetary value in the system is an integer count of the currency's
 * minor unit (cents, fils, sen, …). No function here ever produces or accepts
 * a fractional amount; intermediate products are computed in BigInt so large
 * totals cannot silently lose precision.
 *
 * The Rust implementation (`pos_core::money`) is authoritative for anything
 * persisted. This module exists so the UI can preview identical numbers
 * instantly; both are pinned to `contracts/money-vectors.json`.
 */
import { z } from 'zod';
import { CURRENCIES, type CurrencyCode, type ExchangeRate } from './currency';
import { BPS_SCALE, QUANTITY_SCALE } from './primitives';

/** An integer amount in the minor unit of some currency. */
export type MinorUnits = number;

export const MinorUnitsSchema = z.int();
export const NonNegativeMinorUnitsSchema = z.int().nonnegative();

export const ROUNDING_MODES = ['half_up', 'half_even', 'toward_zero'] as const;
/**
 * - `half_up`: ties round away from zero (common for tax).
 * - `half_even`: ties round to the even neighbour (banker's rounding).
 * - `toward_zero`: truncate.
 */
export type RoundingMode = (typeof ROUNDING_MODES)[number];
export const RoundingModeSchema = z.enum(ROUNDING_MODES);

export class MoneyError extends Error {
  override readonly name = 'MoneyError';
}

export function assertMinorUnits(value: number, label = 'amount'): MinorUnits {
  if (!Number.isSafeInteger(value)) {
    throw new MoneyError(`${label} must be a safe integer in minor units, got ${String(value)}`);
  }
  return value;
}

function toSafeNumber(value: bigint, label: string): MinorUnits {
  if (value > BigInt(Number.MAX_SAFE_INTEGER) || value < BigInt(Number.MIN_SAFE_INTEGER)) {
    throw new MoneyError(`${label} overflowed the safe integer range`);
  }
  return Number(value);
}

function abs(value: bigint): bigint {
  return value < 0n ? -value : value;
}

/** Integer division `numerator / denominator` with an explicit rounding mode. */
export function divRound(numerator: bigint, denominator: bigint, mode: RoundingMode): bigint {
  if (denominator === 0n) throw new MoneyError('division by zero');
  let n = numerator;
  let d = denominator;
  if (d < 0n) {
    n = -n;
    d = -d;
  }
  const quotient = n / d; // BigInt division truncates toward zero
  const remainder = n % d;
  if (remainder === 0n || mode === 'toward_zero') return quotient;

  const step = n < 0n ? -1n : 1n;
  const twiceRemainder = 2n * abs(remainder);
  if (twiceRemainder > d) return quotient + step;
  if (twiceRemainder < d) return quotient;
  // exact tie
  if (mode === 'half_up') return quotient + step;
  return quotient % 2n === 0n ? quotient : quotient + step;
}

export function addMinor(...amounts: readonly MinorUnits[]): MinorUnits {
  let total = 0n;
  for (const amount of amounts) total += BigInt(assertMinorUnits(amount));
  return toSafeNumber(total, 'sum');
}

export function subtractMinor(a: MinorUnits, b: MinorUnits): MinorUnits {
  return toSafeNumber(BigInt(assertMinorUnits(a)) - BigInt(assertMinorUnits(b)), 'difference');
}

/** `unitPrice × quantity`, where quantity is in thousandths (1000 = 1 unit). */
export function multiplyByQuantity(
  unitPrice: MinorUnits,
  quantityMilli: number,
  mode: RoundingMode = 'half_up',
): MinorUnits {
  const product =
    BigInt(assertMinorUnits(unitPrice)) * BigInt(assertMinorUnits(quantityMilli, 'quantity'));
  return toSafeNumber(divRound(product, BigInt(QUANTITY_SCALE), mode), 'line total');
}

/** `amount × bps / 10 000` — percentage discounts, exclusive tax, loyalty earn. */
export function applyBasisPoints(
  amount: MinorUnits,
  bps: number,
  mode: RoundingMode = 'half_up',
): MinorUnits {
  const product = BigInt(assertMinorUnits(amount)) * BigInt(assertMinorUnits(bps, 'basis points'));
  return toSafeNumber(divRound(product, BigInt(BPS_SCALE), mode), 'percentage');
}

/**
 * Splits a tax-inclusive gross amount into net + tax.
 * `net = gross × 10 000 / (10 000 + rate)`, `tax = gross − net`, so
 * `net + tax === gross` always holds.
 */
export function extractInclusiveTax(
  gross: MinorUnits,
  rateBps: number,
  mode: RoundingMode = 'half_up',
): { net: MinorUnits; tax: MinorUnits } {
  const g = BigInt(assertMinorUnits(gross));
  const rate = BigInt(assertMinorUnits(rateBps, 'tax rate'));
  if (rate < 0n) throw new MoneyError('tax rate must be non-negative');
  const scale = BigInt(BPS_SCALE);
  const net = divRound(g * scale, scale + rate, mode);
  return { net: toSafeNumber(net, 'net'), tax: toSafeNumber(g - net, 'tax') };
}

/**
 * Splits `amount` across `weights` so the parts sum exactly to `amount`
 * (largest-remainder method; ties go to the earliest index). Used for split
 * bills and pro-rating order-level discounts across lines.
 */
export function allocate(amount: MinorUnits, weights: readonly number[]): MinorUnits[] {
  const total = BigInt(assertMinorUnits(amount));
  if (weights.length === 0) throw new MoneyError('allocate needs at least one weight');
  const w = weights.map((weight) => {
    const value = BigInt(assertMinorUnits(weight, 'weight'));
    if (value < 0n) throw new MoneyError('weights must be non-negative');
    return value;
  });
  const weightSum = w.reduce((acc, value) => acc + value, 0n);
  if (weightSum === 0n) throw new MoneyError('weights must not all be zero');

  const sign = total < 0n ? -1n : 1n;
  const magnitude = abs(total);
  const parts = w.map((value) => (magnitude * value) / weightSum);
  const remainders = w.map((value, index) => ({ index, rem: (magnitude * value) % weightSum }));
  let leftover = magnitude - parts.reduce((acc, value) => acc + value, 0n);
  remainders.sort((a, b) => (a.rem === b.rem ? a.index - b.index : a.rem > b.rem ? -1 : 1));
  for (const { index } of remainders) {
    if (leftover === 0n) break;
    parts[index] = (parts[index] ?? 0n) + 1n;
    leftover -= 1n;
  }
  return parts.map((part) => toSafeNumber(part * sign, 'allocation'));
}

/**
 * Converts `amount` (minor units of `rate.base`) into minor units of `rate.quote`,
 * accounting for differing exponents (e.g. KWD has 3 decimals, USD has 2).
 */
export function convertCurrency(
  amount: MinorUnits,
  rate: ExchangeRate,
  mode: RoundingMode = 'half_up',
): MinorUnits {
  const fromExp = CURRENCIES[rate.base].exponent;
  const toExp = CURRENCIES[rate.quote].exponent;
  const numerator =
    BigInt(assertMinorUnits(amount)) * BigInt(rate.numerator) * 10n ** BigInt(toExp);
  const denominator = BigInt(rate.denominator) * 10n ** BigInt(fromExp);
  return toSafeNumber(divRound(numerator, denominator, mode), 'converted amount');
}

/** `1500, 'KWD'` → `"1.500"`; `-5, 'USD'` → `"-0.05"`. Exact, no floats. */
export function toDecimalString(amount: MinorUnits, currency: CurrencyCode): string {
  const exponent = CURRENCIES[currency].exponent;
  const negative = assertMinorUnits(amount) < 0;
  const digits = Math.abs(amount).toString();
  if (exponent === 0) return (negative ? '-' : '') + digits;
  const padded = digits.padStart(exponent + 1, '0');
  const major = padded.slice(0, -exponent);
  const minor = padded.slice(-exponent);
  return `${negative ? '-' : ''}${major}.${minor}`;
}

/**
 * Parses user input such as `"1.5"` into minor units (`1500` for KWD).
 * Rejects more fractional digits than the currency allows rather than rounding.
 */
export function parseDecimalString(input: string, currency: CurrencyCode): MinorUnits {
  const exponent = CURRENCIES[currency].exponent;
  const match = /^(-)?(\d+)(?:\.(\d*))?$/.exec(input.trim());
  if (!match) throw new MoneyError(`not a decimal amount: "${input}"`);
  const [, sign, major = '0', fraction = ''] = match;
  if (fraction.length > exponent) {
    throw new MoneyError(`${currency} allows at most ${String(exponent)} decimal places`);
  }
  const minor = BigInt(major + fraction.padEnd(exponent, '0'));
  return toSafeNumber(sign ? -minor : minor, 'parsed amount');
}

/** Locale-aware display, e.g. `formatMoney(1500, 'KWD', 'ar-KW')`. Display only. */
export function formatMoney(amount: MinorUnits, currency: CurrencyCode, locale: string): string {
  const exponent = CURRENCIES[currency].exponent;
  const formatter = new Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
    minimumFractionDigits: exponent,
    maximumFractionDigits: exponent,
  });
  // Intl accepts exact decimal strings, so no float conversion happens here.
  return formatter.format(toDecimalString(amount, currency) as Intl.StringNumericLiteral);
}

File: crates/pos-core/src/money.rs
//! Integer money arithmetic — the authoritative implementation.
//!
//! Amounts are `i64` counts of a currency's minor unit. Intermediate products
//! use `i128`, and every result is checked to fit the JavaScript safe-integer
//! range so it survives the trip over IPC to the UI unchanged.
//!
//! Pinned to `packages/shared/contracts/money-vectors.json` together with the
//! TypeScript twin in `@pos/shared`.

use serde::{Deserialize, Serialize};

use crate::currency::{CurrencyCode, ExchangeRate};

/// An integer amount in the minor unit of some currency.
pub type MinorUnits = i64;

/// 2^53 − 1: the largest integer a JS `number` represents exactly.
pub const JS_MAX_SAFE_INTEGER: i64 = 9_007_199_254_740_991;
/// Quantities are thousandths of a unit: 1000 = 1 unit.
pub const QUANTITY_SCALE: i64 = 1_000;
/// 10 000 basis points = 100 %.
pub const BPS_SCALE: i64 = 10_000;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RoundingMode {
    /// Ties round away from zero.
    HalfUp,
    /// Ties round to the even neighbour (banker's rounding).
    HalfEven,
    /// Truncate.
    TowardZero,
}

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum MoneyError {
    #[error("division by zero")]
    DivisionByZero,
    #[error("{0} overflowed the safe integer range")]
    Overflow(&'static str),
    #[error("{0}")]
    Invalid(String),
}

pub type MoneyResult<T> = Result<T, MoneyError>;

fn to_safe(value: i128, label: &'static str) -> MoneyResult<MinorUnits> {
    if value > i128::from(JS_MAX_SAFE_INTEGER) || value < -i128::from(JS_MAX_SAFE_INTEGER) {
        return Err(MoneyError::Overflow(label));
    }
    i64::try_from(value).map_err(|_| MoneyError::Overflow(label))
}

fn checked(value: Option<i128>, label: &'static str) -> MoneyResult<i128> {
    value.ok_or(MoneyError::Overflow(label))
}

/// Integer division with an explicit rounding mode.
pub fn div_round(numerator: i128, denominator: i128, mode: RoundingMode) -> MoneyResult<i128> {
    if denominator == 0 {
        return Err(MoneyError::DivisionByZero);
    }
    let (n, d) = if denominator < 0 {
        (
            checked(numerator.checked_neg(), "numerator")?,
            checked(denominator.checked_neg(), "denominator")?,
        )
    } else {
        (numerator, denominator)
    };
    let quotient = n / d; // truncates toward zero
    let remainder = n % d;
    if remainder == 0 || mode == RoundingMode::TowardZero {
        return Ok(quotient);
    }
    let step = if n < 0 { -1 } else { 1 };
    let twice_remainder = checked(remainder.abs().checked_mul(2), "remainder")?;
    Ok(match twice_remainder.cmp(&d) {
        std::cmp::Ordering::Greater => quotient + step,
        std::cmp::Ordering::Less => quotient,
        std::cmp::Ordering::Equal => match mode {
            RoundingMode::HalfUp => quotient + step,
            RoundingMode::HalfEven if quotient % 2 == 0 => quotient,
            _ => quotient + step,
        },
    })
}

pub fn add(amounts: &[MinorUnits]) -> MoneyResult<MinorUnits> {
    let total = amounts.iter().map(|a| i128::from(*a)).sum::<i128>();
    to_safe(total, "sum")
}

/// `unit_price × quantity_milli / 1000`.
pub fn multiply_by_quantity(
    unit_price: MinorUnits,
    quantity_milli: i64,
    mode: RoundingMode,
) -> MoneyResult<MinorUnits> {
    let product = i128::from(unit_price) * i128::from(quantity_milli);
    to_safe(
        div_round(product, i128::from(QUANTITY_SCALE), mode)?,
        "line total",
    )
}

/// `amount × bps / 10 000` — percentage discounts, exclusive tax, loyalty earn.
pub fn apply_basis_points(
    amount: MinorUnits,
    bps: i64,
    mode: RoundingMode,
) -> MoneyResult<MinorUnits> {
    let product = i128::from(amount) * i128::from(bps);
    to_safe(
        div_round(product, i128::from(BPS_SCALE), mode)?,
        "percentage",
    )
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub struct TaxSplit {
    pub net: MinorUnits,
    pub tax: MinorUnits,
}

/// Splits a tax-inclusive gross amount; `net + tax == gross` always holds.
pub fn extract_inclusive_tax(
    gross: MinorUnits,
    rate_bps: i64,
    mode: RoundingMode,
) -> MoneyResult<TaxSplit> {
    if rate_bps < 0 {
        return Err(MoneyError::Invalid("tax rate must be non-negative".into()));
    }
    let g = i128::from(gross);
    let scale = i128::from(BPS_SCALE);
    let net = div_round(g * scale, scale + i128::from(rate_bps), mode)?;
    Ok(TaxSplit {
        net: to_safe(net, "net")?,
        tax: to_safe(g - net, "tax")?,
    })
}

/// Splits `amount` across `weights` so the parts sum exactly to `amount`
/// (largest remainder; ties go to the earliest index).
pub fn allocate(amount: MinorUnits, weights: &[i64]) -> MoneyResult<Vec<MinorUnits>> {
    if weights.is_empty() {
        return Err(MoneyError::Invalid(
            "allocate needs at least one weight".into(),
        ));
    }
    if weights.iter().any(|w| *w < 0) {
        return Err(MoneyError::Invalid("weights must be non-negative".into()));
    }
    let weight_sum: i128 = weights.iter().map(|w| i128::from(*w)).sum();
    if weight_sum == 0 {
        return Err(MoneyError::Invalid("weights must not all be zero".into()));
    }

    let sign: i128 = if amount < 0 { -1 } else { 1 };
    let magnitude = i128::from(amount).abs();
    let mut parts: Vec<i128> = weights
        .iter()
        .map(|w| magnitude * i128::from(*w) / weight_sum)
        .collect();
    let mut remainders: Vec<(usize, i128)> = weights
        .iter()
        .enumerate()
        .map(|(i, w)| (i, magnitude * i128::from(*w) % weight_sum))
        .collect();
    let mut leftover = magnitude - parts.iter().sum::<i128>();
    remainders.sort_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));
    for (index, _) in remainders {
        if leftover == 0 {
            break;
        }
        parts[index] += 1;
        leftover -= 1;
    }
    parts
        .into_iter()
        .map(|p| to_safe(p * sign, "allocation"))
        .collect()
}

/// Converts minor units of `rate.base` into minor units of `rate.quote`.
pub fn convert_currency(
    amount: MinorUnits,
    rate: &ExchangeRate,
    mode: RoundingMode,
) -> MoneyResult<MinorUnits> {
    if rate.numerator == 0 || rate.denominator == 0 {
        return Err(MoneyError::Invalid("exchange rate must be positive".into()));
    }
    let to_scale = 10_i128.pow(rate.quote.exponent());
    let from_scale = 10_i128.pow(rate.base.exponent());
    let numerator = checked(
        i128::from(amount)
            .checked_mul(i128::from(rate.numerator))
            .and_then(|v| v.checked_mul(to_scale)),
        "converted amount",
    )?;
    let denominator = checked(
        i128::from(rate.denominator).checked_mul(from_scale),
        "converted amount",
    )?;
    to_safe(div_round(numerator, denominator, mode)?, "converted amount")
}

/// `1500, KWD` → `"1.500"`. Exact, no floats.
pub fn to_decimal_string(amount: MinorUnits, currency: CurrencyCode) -> String {
    let exponent = currency.exponent() as usize;
    let digits = amount.unsigned_abs().to_string();
    let sign = if amount < 0 { "-" } else { "" };
    if exponent == 0 {
        return format!("{sign}{digits}");
    }
    let padded = format!("{digits:0>width$}", width = exponent + 1);
    let (major, minor) = padded.split_at(padded.len() - exponent);
    format!("{sign}{major}.{minor}")
}

/// Parses `"1.5"` into minor units (`1500` for KWD). Rejects excess precision.
pub fn parse_decimal_string(input: &str, currency: CurrencyCode) -> MoneyResult<MinorUnits> {
    let invalid = || MoneyError::Invalid(format!("not a decimal amount: \"{input}\""));
    let exponent = currency.exponent() as usize;
    let trimmed = input.trim();
    let (negative, unsigned) = match trimmed.strip_prefix('-') {
        Some(rest) => (true, rest),
        None => (false, trimmed),
    };
    let (major, fraction) = match unsigned.split_once('.') {
        Some((major, fraction)) => (major, fraction),
        None => (unsigned, ""),
    };
    let all_digits = |s: &str| s.bytes().all(|b| b.is_ascii_digit());
    if major.is_empty() || !all_digits(major) || !all_digits(fraction) {
        return Err(invalid());
    }
    if fraction.len() > exponent {
        return Err(MoneyError::Invalid(format!(
            "{} allows at most {exponent} decimal places",
            currency.as_str()
        )));
    }
    let combined = format!("{major}{fraction:0<exponent$}");
    let magnitude: i128 = combined.parse().map_err(|_| invalid())?;
    to_safe(
        if negative { -magnitude } else { magnitude },
        "parsed amount",
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn overflow_is_detected() {
        assert_eq!(
            multiply_by_quantity(JS_MAX_SAFE_INTEGER, 2_000, RoundingMode::HalfUp),
            Err(MoneyError::Overflow("line total"))
        );
    }

    #[test]
    fn division_by_zero_is_an_error() {
        assert_eq!(
            div_round(1, 0, RoundingMode::HalfUp),
            Err(MoneyError::DivisionByZero)
        );
    }
}
